sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala

```diff
@@ -28,12 +28,18 @@ class SparkOptimizer(
     experimentalMethods: ExperimentalMethods)
   extends Optimizer(catalog) {
 
-  override def batches: Seq[Batch] = (preOptimizationBatches ++ super.batches :+
+  val experimentalPreOptimizations: Seq[Batch] = Seq(Batch(
+    "User Provided Pre Optimizers", fixedPoint, experimentalMethods.extraPreOptimizations: _*))
+
+  val experimentalPostOptimizations: Batch = Batch(
+    "User Provided Post Optimizers", fixedPoint, experimentalMethods.extraOptimizations: _*)
+
+  override def batches: Seq[Batch] = experimentalPreOptimizations ++
+    (preOptimizationBatches ++ super.batches :+
```

Comment:
why can't user just use preOptimizationBatches?

Reply:
@wzhfy Thanks for your comment. Yes, i see preOptimizationBatches is introduced since 2.2 but i'm not sure this option allows user to add custom rules during runtime (say, via spark-shell). Could you confirm this? Thanks.

Reply:
For example, there is `postHocOptimizationBatches` but `experimentalMethods.extraOptimizations` is used for adding custom optimisation methods..

Reply:
OK, I see. Then could you add the use case to PR description? like: 
```
after this PR, we can add both pre/post optimization rules at runtime as follows:
...
```

Reply:
This PR is not about Analyzer, please also update your description.